src/parser.rs

```diff
@@ -279,61 +334,62 @@ named!(relational_expression<CompleteStr, RelationalExpression>,
         space0 >>
         relation: equality_relation >>
         space0 >>
-        right_expression: string_expression >>
+        right_expression: return_error!(ErrorKind::Custom(
+            ErrorCode::ExpectedStringExpression as u32), string_expression) >>
```

Comment:
What is `return_error!` doing?

Reply:
return_error is `nom`'s macro, which terminates evaluation immediately. In nom you can enable a feature to have rich error types, then it will capture a lot of context, but also do more work and allocations. This default feature allows you just to use a simple error code number, and nothing else. So, this is what I am doing here. I defined an enum for error numbers and just cast it to an error code which is u32.

Reply:
So it returns an error if `right_expression` is not a `string_expression`. Did I understand this right?